modules/core/src/test/java/org/apache/ignite/failure/TestFailingFailureHandler.java

```diff
@@ -0,0 +1,62 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one or more
+ * contributor license agreements.  See the NOTICE file distributed with
+ * this work for additional information regarding copyright ownership.
+ * The ASF licenses this file to You under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License.  You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.ignite.failure;
+
+import junit.framework.TestCase;
+import org.apache.ignite.Ignite;
+import org.apache.ignite.internal.util.tostring.GridToStringExclude;
+import org.apache.ignite.internal.util.typedef.internal.S;
+import org.apache.ignite.testframework.junits.GridAbstractTest;
+
+import static org.apache.ignite.testframework.junits.GridAbstractTest.testIsRunning;
+
+/**
+ * Stops node and fails test.
+ */
+public class TestFailingFailureHandler extends StopNodeFailureHandler {
+    /** Test. */
+    @GridToStringExclude
+    protected final GridAbstractTest test;
+
+    /**
+     * @param test Test.
+     */
+    public TestFailingFailureHandler(GridAbstractTest test) {
+        this.test = test;
+    }
+
+    /** {@inheritDoc} */
+    @Override public boolean handle(Ignite ignite, FailureContext failureCtx) {
+        if (!testIsRunning.getAndSet(false)) {
```

Comment:
Is it possible that a failure handler will be called by different threads with different exceptions? Will these exceptions be visible in a resulting test failure?

Reply:
Yes, they will be visible, because in most cases error is printed before moving to failure handler. In other cases error is thrown to the upper code, where it is logged with additional actions.